diff -Npru --exclude='.git*' AOSP/build/core/pathmap.mk pdroid/build/core/pathmap.mk
--- AOSP/build/core/pathmap.mk	2012-07-13 11:05:16.605001060 -0700
+++ pdroid/build/core/pathmap.mk	2012-07-13 14:59:32.459835410 -0700
@@ -84,6 +84,7 @@ JNI_H_INCLUDE := $(call include-path-for
 # is so common and the other framework libraries can have mutual
 # interdependencies.
 #
+# BEGIN privacy-added: privacy
 FRAMEWORKS_BASE_SUBDIRS := \
 	$(addsuffix /java, \
 	    core \
@@ -95,6 +96,7 @@ FRAMEWORKS_BASE_SUBDIRS := \
 	    media/mca/filterpacks \
 	    drm \
 	    opengl \
+	    privacy \
 	    sax \
 	    telephony \
 	    wifi \
@@ -102,6 +104,7 @@ FRAMEWORKS_BASE_SUBDIRS := \
 	    icu4j \
 	    voip \
 	 )
+# END privacy-added: privacy
 
 #
 # A version of FRAMEWORKS_BASE_SUBDIRS that is expanded to full paths from
diff -Npru --exclude='.git*' AOSP/build/target/product/core.mk pdroid/build/target/product/core.mk
--- AOSP/build/target/product/core.mk	2012-07-13 11:05:16.622005094 -0700
+++ pdroid/build/target/product/core.mk	2012-07-13 14:59:32.459835410 -0700
@@ -22,6 +22,7 @@ PRODUCT_PROPERTY_OVERRIDES := \
     ro.config.notification_sound=OnTheHunt.ogg \
     ro.config.alarm_alert=Alarm_Classic.ogg
 
+# BEGIN privacy-added: PDroid
 PRODUCT_PACKAGES := \
     ApplicationsProvider \
     BackupRestoreConfirmation \
@@ -35,6 +36,7 @@ PRODUCT_PACKAGES := \
     Home \
     KeyChain \
     MediaProvider \
+    PDroid \
     PackageInstaller \
     PicoTts \
     SettingsProvider \
@@ -120,6 +122,7 @@ PRODUCT_PACKAGES := \
     sensorservice \
     lint \
     uiautomator
+# END privacy-added: PDroid
 
 PRODUCT_COPY_FILES += \
     system/core/rootdir/init.usb.rc:root/init.usb.rc \
diff -Npru --exclude='.git*' AOSP/build/target/product/security/privacy.x509.pem pdroid/build/target/product/security/privacy.x509.pem
--- AOSP/build/target/product/security/privacy.x509.pem	1969-12-31 16:00:00.000000000 -0800
+++ pdroid/build/target/product/security/privacy.x509.pem	2012-07-13 14:59:32.524850920 -0700
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
+MIIDNDCCAhwCCQCXWkdOAj3rWTANBgkqhkiG9w0BAQUFADBcMQswCQYDVQQGEwJE
+RTETMBEGA1UECBMKU29tZS1TdGF0ZTEhMB8GA1UEChMYSW50ZXJuZXQgV2lkZ2l0
+cyBQdHkgTHRkMRUwEwYDVQQDEwxTdnlhdG9zbGF2IEgwHhcNMTEwNzE0MjEyNTI2
+WhcNMzgxMTI4MjEyNTI2WjBcMQswCQYDVQQGEwJERTETMBEGA1UECBMKU29tZS1T
+dGF0ZTEhMB8GA1UEChMYSW50ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMRUwEwYDVQQD
+EwxTdnlhdG9zbGF2IEgwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDb
+i4kL71TqNRi0F5RjDVejPlnN1aiFtruvWiuvppcPjY9SNw6WaTfJS356yWBrorU6
+qZbqS0l5vl+CcMrm94F2gbTZamsHaUz7puGMLx3R/s3vvYSD8pPwzqhyLY+mSSK+
+cbhvthJMa4FJLFyHKlHj7l1wn8qeAr2ceKVF4w5SjwGl9hwSHA68TMeb9dWO87N7
+82p+D9fqU3I33Qf0ieiBiRYrEBHQtDwCMJNktNmmG9GyKyu8PVs2jhvJkNsga/Le
+y+0FPwRgKA0MgnKavliGxST76Sl98pimc9h+ZyedwXw2z+lk1Fh+qhuDBWaTgIJl
+C+yLqa+TAYqimUdUYSuvAgMBAAEwDQYJKoZIhvcNAQEFBQADggEBAHI5T9gdsFh1
+X4g7eQVXBEyW11mpsqUh/Kl4PI/8xSzeon9DQcNc+oQJDAvtOoQXww1sYKUykmwr
+H7idjx6fMyygeUsvRy7clDBfdG+GR43QPkJSZT61ZGhZUbKKOPoAMnomNe2uL4LW
+4iU/1bkEQNKzV0iea663t3t5rr++lE2Ulm0LIDLgxlmuyvl/6e8kgRk/ojj/ZhIh
+IjEZjWwAJDRxGz7Kc19keecUtIE/fBZrkC+qRivnfqEuXH/MW7a18uYYtG25yFQp
+UJHQ0wBayhY9NKkcgVAABMdiNtKJIXdrkeKBKDSnuX7GErwy23adjZ5wFjxefm7P
+jjazBHHW1qc=
+-----END CERTIFICATE-----
