diff --git a/core/pathmap.mk b/core/pathmap.mk
index 5dfc543..6b24803 100644
--- a/core/pathmap.mk
+++ b/core/pathmap.mk
@@ -86,6 +86,7 @@ FRAMEWORKS_BASE_SUBDIRS := \
 	    media/mca/filterpacks \
 	    drm \
 	    opengl \
+	    privacy \
 	    sax \
 	    telephony \
 	    wifi \
diff --git a/target/product/core_minimal.mk b/target/product/core_minimal.mk
index fc2fc80..3538702 100644
--- a/target/product/core_minimal.mk
+++ b/target/product/core_minimal.mk
@@ -28,6 +28,7 @@ PRODUCT_PACKAGES += \
     HTMLViewer \
     MediaProvider \
     PackageInstaller \
+    PDroid \
     SettingsProvider \
     Shell \
     bu \
